File: e2e/factories/equityGrants.ts
import { db, takeOrThrow } from "@test/db";
import { companyInvestorEntitiesFactory } from "@test/factories/companyInvestorEntities";
import { companyInvestorsFactory } from "@test/factories/companyInvestors";
import { optionPoolsFactory } from "@test/factories/optionPools";
import { addMonths, endOfDay, startOfDay, subDays } from "date-fns";
import { and, eq } from "drizzle-orm";
import { companyInvestorEntities, companyInvestors, equityGrants, optionPools, users } from "@/db/schema";
import { assert } from "@/utils/assert";

let grantCounter = 0;

type CreateOptions = {
  year: number;
};

const defaultOptions = { year: new Date().getFullYear() - 2 };

export const equityGrantsFactory = {
  create: async (
    overrides: Partial<typeof equityGrants.$inferInsert> = {},
    options: CreateOptions = defaultOptions,
  ) => {
    const investor = overrides.companyInvestorId
      ? await db.query.companyInvestors
          .findFirst({
            where: eq(companyInvestors.id, overrides.companyInvestorId),
            with: { user: true },
          })
          .then(takeOrThrow)
      : (await companyInvestorsFactory.create()).companyInvestor;

    const investorUser = await db.query.users
      .findFirst({
        where: eq(users.id, investor.userId),
      })
      .then(takeOrThrow);

    // companyInvestorEntity should have the same company and legalName as the investor
    const companyInvestorEntity = overrides.companyInvestorEntityId
      ? await db.query.companyInvestorEntities
          .findFirst({
            where: and(
              eq(companyInvestorEntities.id, overrides.companyInvestorEntityId),
              eq(companyInvestorEntities.companyId, investor.companyId),
              eq(companyInvestorEntities.name, investorUser.legalName ?? ""),
            ),
          })
          .then(takeOrThrow)
      : (
          await companyInvestorEntitiesFactory.create({
            companyId: investor.companyId,
            name: investorUser.legalName ?? "",
          })
        ).companyInvestorEntity;
    const optionPool = overrides.optionPoolId
      ? await db.query.optionPools
          .findFirst({
            where: and(eq(optionPools.id, overrides.optionPoolId), eq(optionPools.companyId, investor.companyId)),
          })
          .then(takeOrThrow)
      : (
          await optionPoolsFactory.create({
            companyId: investor.companyId,
          })
        ).optionPool;

    grantCounter++;

    const [equityGrant] = await db
      .insert(equityGrants)
      .values({
        companyInvestorId: investor.id,
        companyInvestorEntityId: companyInvestorEntity.id,
        optionPoolId: optionPool.id,
        name: overrides.name || `GUM-${grantCounter}`,
        numberOfShares: overrides.numberOfShares || 100,
        sharePriceUsd: overrides.sharePriceUsd || "10",
        exercisePriceUsd: overrides.exercisePriceUsd || "5",
        vestedShares: overrides.vestedShares || 100,
        unvestedShares: overrides.unvestedShares || 0,
        vestingTrigger: overrides.vestingTrigger || "invoice_paid",
        exercisedShares: overrides.exercisedShares || 0,
        forfeitedShares: overrides.forfeitedShares || 0,
        issuedAt: overrides.issuedAt || new Date(),
        expiresAt: overrides.expiresAt || addMonths(new Date(), optionPool.defaultOptionExpiryMonths),
        acceptedAt: overrides.acceptedAt || new Date(),
        optionHolderName: overrides.optionHolderName || (investorUser.legalName ?? ""),
        boardApprovalDate: overrides.boardApprovalDate || subDays(new Date(), 1).toDateString(),
        voluntaryTerminationExerciseMonths: overrides.voluntaryTerminationExerciseMonths || 120,
        involuntaryTerminationExerciseMonths: overrides.involuntaryTerminationExerciseMonths || 120,
        terminationWithCauseExerciseMonths: overrides.terminationWithCauseExerciseMonths || 0,
        deathExerciseMonths: overrides.deathExerciseMonths || 120,
        disabilityExerciseMonths: overrides.disabilityExerciseMonths || 120,
        retirementExerciseMonths: overrides.retirementExerciseMonths || 120,
        periodStartedAt: overrides.periodStartedAt || startOfDay(new Date(`${options.year}-01-01`)),
        periodEndedAt: overrides.periodEndedAt || endOfDay(new Date(`${options.year}-12-31`)),
        ...overrides,
      })
      .returning();
    assert(equityGrant != null);

    return { equityGrant };
  },

  createActive: async (
    overrides: Partial<typeof equityGrants.$inferInsert> = {},
    options: CreateOptions = defaultOptions,
  ) =>
    equityGrantsFactory.create(
      {
        numberOfShares: 1000,
        vestedShares: 100,
        unvestedShares: 700,
        exercisedShares: 200,
        forfeitedShares: 0,
        ...overrides,
      },
      options,
    ),
};

File: backend/spec/factories/equity_grants.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :equity_grant do
    company_investor
    company_investor_entity do
      association :company_investor_entity,
                  company: company_investor.company,
                  name: company_investor.user.legal_name
    end
    option_pool do
      association :option_pool,
                  company: company_investor.company
    end
    sequence(:name) { |n| "GUM-#{n}" }
    number_of_shares { 100 }
    share_price_usd { 10 }
    exercise_price_usd { 5 }
    vested_shares { number_of_shares }
    unvested_shares { 0 }
    vesting_trigger { :invoice_paid }
    exercised_shares { 0 }
    forfeited_shares { 0 }
    issued_at { Time.current }
    expires_at { issued_at + option_pool.default_option_expiry_months.months }
    accepted_at { DateTime.current }
    option_holder_name { company_investor.user.legal_name }
    board_approval_date { Date.today - 1.day }
    voluntary_termination_exercise_months { 120 }
    involuntary_termination_exercise_months { 120 }
    termination_with_cause_exercise_months { 0 }
    death_exercise_months { 120 }
    disability_exercise_months { 120 }
    retirement_exercise_months { 120 }
    transient do
      year { DateTime.current.year - 2 }
    end
    after(:build) do |grant, evaluator|
      grant.period_started_at ||= DateTime.parse("1 Jan #{evaluator.year}")
      grant.period_ended_at ||= DateTime.parse("31 Dec #{evaluator.year}")
    end

    factory :active_grant do
      number_of_shares { 1000 }
      vested_shares { 100 }
      unvested_shares { 700 }
      exercised_shares { 200 }
      forfeited_shares { 0 }
    end


    trait :vests_on_invoice_payment do
      vested_shares { 0 }
      unvested_shares { number_of_shares }
      vesting_trigger { :invoice_paid }

      after(:create) do |grant|
        grant.build_vesting_events.each(&:save!)
      end
    end

    trait :vests_as_per_schedule do
      vesting_trigger { :scheduled }
      vested_shares { 0 }
      unvested_shares { number_of_shares }
      vesting_schedule
      period_started_at { board_approval_date.beginning_of_day }
      period_ended_at { period_started_at.end_of_day + vesting_schedule.total_vesting_duration_months.months }

      after(:create) do |grant|
        grant.build_vesting_events.each(&:save!)
      end
    end

    # So that virtual attributes are calculated correctly
    after(:create) {  _1.reload }
  end
end
